backend/internal/oauth/oauth2/authz/utils.go

```diff
@@ -0,0 +1,270 @@
+	authzmodel "github.com/asgardeo/thunder/internal/oauth/oauth2/authz/model"
+	oauth2const "github.com/asgardeo/thunder/internal/oauth/oauth2/constants"
+	sessionstore "github.com/asgardeo/thunder/internal/oauth/session/store"
+	"github.com/asgardeo/thunder/internal/system/config"
+	jwtutils "github.com/asgardeo/thunder/internal/system/crypto/jwt/utils"
+	"github.com/asgardeo/thunder/internal/system/log"
+	"github.com/asgardeo/thunder/internal/system/utils"
+	systemutils "github.com/asgardeo/thunder/internal/system/utils"
+)
+
+// getOAuthMessage extracts the OAuth message from the request and response writer.
+func getOAuthMessage(r *http.Request, w http.ResponseWriter) *authzmodel.OAuthMessage {
+	logger := log.GetLogger()
+	if r == nil || w == nil {
+		logger.Error("Request or response writer is nil")
+		return nil
+	}
+
+	var msg *authzmodel.OAuthMessage
+	var err error
+
+	switch r.Method {
+	case http.MethodGet:
+		msg, err = getOAuthMessageForGetRequest(r)
+	case http.MethodPost:
+		msg, err = getOAuthMessageForPostRequest(r)
+	default:
+		err = errors.New("unsupported request method: " + r.Method)
+	}
+
+	if err != nil {
+		utils.WriteJSONError(w, oauth2const.ErrorInvalidRequest, "Invalid authorization request",
+			http.StatusBadRequest, nil)
+	}
+
+	return msg
+}
+
+// getOAuthMessageForGetRequest extracts the OAuth message from a authorization GET request.
+func getOAuthMessageForGetRequest(r *http.Request) (*authzmodel.OAuthMessage, error) {
+	if err := r.ParseForm(); err != nil {
+		return nil, errors.New("failed to parse form data: " + err.Error())
+	}
+
+	queryParams := make(map[string]string)
+	for key, values := range r.URL.Query() {
+		if len(values) > 0 {
+			queryParams[key] = values[0]
+		}
+	}
+
+	return &authzmodel.OAuthMessage{
+		RequestType:        oauth2const.TypeInitialAuthorizationRequest,
+		SessionData:        nil,
+		RequestQueryParams: queryParams,
+	}, nil
+}
+
+// getOAuthMessageForPostRequest extracts the OAuth message from a authorization POST request.
+func getOAuthMessageForPostRequest(r *http.Request) (*authzmodel.OAuthMessage, error) {
+	authZReq, err := systemutils.DecodeJSONBody[authzmodel.AuthZPostRequest](r)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode JSON body: %w", err)
+	}
+
+	if authZReq.SessionDataKey == "" || authZReq.Assertion == "" {
+		return nil, errors.New("sessionDataKey or assertion is missing")
+	}
+
+	// Determine the request type.
+	// TODO: Require to handle other types such as user consent, etc.
+	requestType := oauth2const.TypeAuthorizationResponseFromEngine
+
+	sessionDataKey := authZReq.SessionDataKey
+	sessionDataStore := sessionstore.GetSessionDataStore()
+	ok, sessionData := sessionDataStore.GetSession(sessionDataKey)
+	if !ok {
+		return nil, fmt.Errorf("session data not found for session data key: %s", sessionDataKey)
+	}
+
+	// Remove the session data after retrieval.
+	sessionDataStore.ClearSession(sessionDataKey)
```

Comment:
An an improvement, we better to avoid using generalized SessionDataStore and have more specific storage implementation.

Reply:
Will track as an improvement

Reply:
Created an issue to track https://github.com/asgardeo/thunder/issues/255